frontend/src/AppBuilder/_hooks/useExposeVariables.js

```diff
@@ -0,0 +1,99 @@
+import { useEffect, useState, useCallback } from 'react';
+
+export const useExposeState = (loadingState, visibleState, disabledState, setExposedVariables, setExposedVariable) => {
+  const [isDisabled, setDisable] = useState(loadingState || disabledState);
+  const [isVisible, setVisibility] = useState(visibleState || true);
+  const [isLoading, setLoading] = useState(loadingState || false);
+
+  // Wrapping state update functions with useCallback prevent rerenders
+  const setDisableIfChangedWithSideEffect = useCallback(
+    (newValue) => {
+      setDisable((prev) => {
+        if (prev !== newValue) {
+          setExposedVariable('isDisabled', newValue);
+          return newValue;
+        }
+        return prev;
+      });
+    },
+    [setExposedVariable]
+  );
+
+  const setVisibilityIfChangedWithSideEffect = useCallback(
+    (newValue) => {
+      setVisibility((prev) => {
+        if (prev !== newValue) {
+          setExposedVariable('isVisible', newValue);
+          return newValue;
+        }
+        return prev;
+      });
+    },
```

Comment:
This logic might affect some of the existing functionality. For example, at present we're setting the value of the component and then we're exposing the value. In this logic, we're trying to expose the value before setting it to the component.

Reply:
Understood. I'll make the changes